### PR TITLE
🗑 Delete message UI

### LINKED ISSUE #12286: Delete messages
As a user
I want to be able to delete messages
So that I can correct any potential mistake I made

### Description

We should provide users a way to delete messages, that should delete messages on both devices.
We should keep the data on the device in order to show a record. 
The implementation can be done similarly to how we did with edit messages.

### PR DESCRIPTION
**Pull request by shivekkhurana**
fixes #12286 

### Summary
Add the option to delete chat message. Also introduce a new UI for message actions.

<details>
  <summary>View screenshot 📸</summary>
  
Simulator Screen Shot - iPhone 11 - 2021-08-03 at 10 29 33

</details>

### Review notes
Messages are critical in terms of perf. Verify that this PR doesn't introduces unintended pitfalls.

#### Platforms
- Android
- iOS

#### Areas that maybe impacted
- all chats
 
##### Functional

- 1-1 chats
- public chats
- group chats

### Steps to test
- Send a message (public or private)
- Long press on a message sent by you to see message actions
- You should see a "Delete" option along with other options
- Clicking delete hides the message in the db
- Deleting doesn't affect the receiver's database

status: WIP

### CODE REVIEW

**Comment by flexsurfer** [comment]
@shivekkhurana is it wip ?is draft

**Comment by shivekkhurana** [comment]
@flexsurfer  It's wip. I'm not clear if the messages will be deleted from the network, and if yes, how to handle them. Started a conversation thread on Discord [here](https://discord.com/channels/624205794384281629/862717376915308544/869891381424959518).

**Comment by shivekkhurana** [comment]
@flexsurfer I have implemented a UI only message delete. It doesn't rebuilds the list on every delete.
Will update other instances of `hide-message` when this gets merged.

**Comment by shivekkhurana** [comment]
**UPDATE : Resolved**

Known Bug: Deleted message is visible to the receiver until they go-back to the chat index and open the chat again.

> https://discord.com/channels/624205794384281629/862717376915308544/872026952754663466
> While implementing the delete-message UI, I realised that the signal doesn't have the id of the deleted message.
> Imagine this flow: D1 and D2 are two devices with a 1:1 chat open.
> 1. D1 sends the message "2" to D2
> 2. D1 sends the message "4" to D2
> 3. D1 deletes the message "4"
> 
> At step 3, D2 receives a signal like this:
```edn
{:description "",
 :lastClockValue 1627977236319,
 :deletedAtClockValue 0,
 :joined 1627976672860,
 :identicon "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADIAAAAyCAYAAAAeP4ixAAAAmklEQVR4nOzY0QnCQBAGYRV7sSnLsAzLsCmrSd6PHLnlODL8zPcoizAsLEcetxCG0BhCYwiNITQxIc/Rwe//tx39/nm971fMt2I2YgiNIVokZiOG0BhCM/3W6qm+qar/04rZiCE0htAMX62e0atyNl+9Zq2YjRhCYwjN9NXqmf1OVRWzEUNoDNEiMRsxhMYQGkNoDKHZAwAA//8uLB1h/9556wAAAABJRU5ErkJggg==",
 :color "",
 :name "0x045238",
 :alias "Remarkable Uniform Nudibranch",
 :lastMessage {:messageType 1,
               :identicon "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADIAAAAyCAYAAAAeP4ixAAAAoUlEQVR4nOzYwQnCQBBGYRV7sSQbsQwbsSSrSe4hC7PZXXj8vO88Bx8Dg5vHLYQhNIbQGEJjCE1MyLM6+P3/trU/5dzn9b5X5mI2YgiNITSli3BF68pVr1CvmI0YQmMIzfAFmfUfbPSaxWzEEBpDaMovxF6tK7TqpRmzEUNoDKEZ/q7Ve51mzR/FbMQQGkO0SMxGDKExhMYQGkNo9gAAAP//YnAcZw9HxUQAAAAASUVORK5CYII=",
               :replace "",
               :chatId "0x0452380654d8e40fcd056db49b1c71c99358d5bc320b1ed0b862c400cbf3201ed7cfb1a480b245c69614453381e5bff2fd8ccffcafb4b9155883f58ef2511d95a9",
               :lineCount 0,
               :whisperTimestamp 1627977123593,
               :alias "Welloff Impressive Reptile",
               :seen true,
               :gapParameters {},
               :from "0x040f2f9a3ccba626625e4479fe61a806dca36fcb453050c21adc36f6d8f024611869326464d6143e2339a9ba7f3f477abe6a9e56fb94cd46b5df14378f50e43d43",
               :id "0x6214e4b09359d6fa73f0fcc302b1089be603676a1fd1a5a147f8d885e42ab0f6",
               :parsedText [{:type "paragraph",
                             :children [{:literal "2"}]}],
               :contentType 1,
               :clock 1627977123593,
               :localChatId "0x0452380654d8e40fcd056db49b1c71c99358d5bc320b1ed0b862c400cbf3201ed7cfb1a480b245c69614453381e5bff2fd8ccffcafb4b9155883f58ef2511d95a9",
               :outgoingStatus "delivered",
               :timestamp 1627977123593,
               :ensName "",
               :quotedMessage nil,
               :rtl false,
               :responseTo "",
               :text "2"},
 :active true,
 :id "0x0452380654d8e40fcd056db49b1c71c99358d5bc320b1ed0b862c400cbf3201ed7cfb1a480b245c69614453381e5bff2fd8ccffcafb4b9155883f58ef2511d95a9",
 :unviewedMentionsCount 0,
 :unviewedMessagesCount 0,
 :membershipUpdateEvents nil,
 :chatType 1,
 :members nil,
 :timestamp 1627977237846}
```
> This signal has no reference to the deleted message.  There is no problem if the chat is in the background or the user is on some other screen. But if the user is on the chat screen, I need some id to remove the deleted message from the screen.

File: src/status_im/chat/models/input.cljs [hunk 1]
```diff
@@ -134,6 +134,33 @@
              (update-in [:chat/inputs current-chat-id :metadata]
                         dissoc :sending-image))}))
 
+(fx/defn soft-delete-message-success
+  {:events [::soft-delete-message-success]}
+  [{:keys [db]} chat-id message-id]
+  (let [new-last-message (-> db
```
Comment (flexsurfer) [inline src/status_im/chat/models/input.cljs]: why doesn't `wakuext_deleteMessageAndSend` return updated chats with updated last message ?

Comment (flexsurfer) [reply src/status_im/chat/models/input.cljs]: i mean it doesn't look correct, to set last message in status-react

File: src/status_im/ui/screens/chat/message/message.cljs [hunk 2]
```diff
@@ -549,25 +560,26 @@
...
-                        (re-frame/dispatch [::models.reactions/send-emoji-reaction-retraction
-                                            {:message-id        (:message-id message)
-                                             :emoji-id          emoji-id
-                                             :emoji-reaction-id emoji-reaction-id}]))
-     :render          ->message}]
-   (when pinned
-     [react/view {:style (style/pin-indicator-container outgoing)}
-      [pinned-by-indicator outgoing display-photo? pinned-by]])])
+(defn chat-message [{:keys [outgoing display-photo? pinned pinned-by hidden-in-ui?] :as message} space-keeper]
+  (when-not hidden-in-ui?
```
Comment (flexsurfer) [inline src/status_im/ui/screens/chat/message/message.cljs]: why don't we filter it in the subscription?

File: src/status_im/chat/models/input.cljs [hunk 3]
```diff
@@ -134,6 +134,33 @@
...
+                             :messages
+                             (get chat-id)
+                             vals
+                             (->> (sort-by :whisper-timestamp)
+                                  (filter #(and (not (:hidden-in-ui? %))
```
Comment (cammellos) [inline src/status_im/chat/models/input.cljs]: I think  you might not need a `hidden-in-ui?` field, we have already `deleted` I believe, and there's a flag we can expose as well to the JSON `hide`.
Also I think the logic of replacing the last message should go in status-go, we should read the response from status go (MessengerResponse) and just process that as we would with any signal.

**Comment by churik** [comment]
@shivekkhurana 
please ping me once it is ready to be tested.
Thank you.

File: src/status_im/ui/screens/chat/message/message.cljs (hunk 2 above)
Comment (shivekkhurana) [reply src/status_im/ui/screens/chat/message/message.cljs]: Got rid of this.

File: src/status_im/chat/models/input.cljs (hunk 3 above)
Comment (shivekkhurana) [reply src/status_im/chat/models/input.cljs]: Got rid of `hidden in ui` will see how to move login to status-go.

File: src/status_im/chat/models/input.cljs (hunk 1 above)
Comment (shivekkhurana) [reply src/status_im/chat/models/input.cljs]: https://github.com/status-im/status-react/pull/12390#discussion_r683900075

**Comment by shivekkhurana** [comment]
@churik This is ready to test

File: src/status_im/chat/models/input.cljs [hunk 4]
```diff
@@ -6,6 +6,7 @@
             [status-im.chat.constants :as chat.constants]
             [status-im.ethereum.json-rpc :as json-rpc]
             [status-im.chat.models :as chat]
+            [status-im.utils.debounce :as debounce]
```
Comment (flexsurfer) [inline src/status_im/chat/models/input.cljs]: isn't used?

File: src/status_im/signals/core.cljs [hunk 5]
```diff
@@ -52,6 +52,7 @@
   (let [^js data (.parse js/JSON event-str)
         ^js event-js (.-event data)
         type (.-type data)]
+    (prn :---process-> type)
```
Comment (flexsurfer) [inline src/status_im/signals/core.cljs]: leftover

File: src/status_im/chat/models/message.cljs [hunk 6]
```diff
@@ -194,3 +194,46 @@
 (fx/defn send-messages
   [cofx messages]
   (protocol/send-chat-messages cofx messages))
+
+(defn flip-args [f]
+  (fn [x y] (f y x)))
+
+(defn message-ids->message-id-chat-id-map
```
Comment (flexsurfer) [inline src/status_im/chat/models/message.cljs]: mhm , why is it so complex ? isn't it possible to have this in the format we want from status-go ?

Comment (shivekkhurana) [reply src/status_im/chat/models/message.cljs]: Currently, the backend only returns the `message-id` of the deleted messages. But messages are indexed by `chat-id` in the frontend model. Its not ideal. I'll check if I can re-write the backend. There is one more issue that needs to be solved.

File: src/status_im/chat/models/input.cljs (hunk 4 above)
Comment (shivekkhurana) [reply src/status_im/chat/models/input.cljs]: yeah, was trying something. This is not needed.

File: src/status_im/chat/models/message.cljs (hunk 6 above)
Comment (flexsurfer) [reply src/status_im/chat/models/message.cljs]: i mean its not urgent, i think thats fine

**Review by flexsurfer — APPROVED** [review]

**Comment by qoqobolo** [comment]
@shivekkhurana 

### ISSUE 1: Recipient sees the deleted message if it was received while Recipient was offline/logged out.
**_Reproduction:_**
1. D1 starts or join any chat with D2 (1-1/group/public/community)
2. D2 closes Status
3. D1 sends a message and deletes it
4. D2 logs in and opens the chat

**Expected behaviour:** no message in chat history
**Actual behaviour:** message is visible

### ISSUE 2: Unread indicator/counter is shown for deleted message 
**_Reproduction:_**
1. D1 and D2 join the same public chat
2. D1 navigates to Home screen
3. D2 sends a message and deletes it

**Expected behaviour:** unread indicator disappears after deleting a message
**Actual behaviour:** indicator remains on chat preview

https://user-images.githubusercontent.com/67952253/128861031-60a97612-6c6f-45d8-a376-d7d466fc955f.mp4

### ISSUE 3: Deleted message is visible in Activity center

If D1 sends a message (in 1-1 or mention/reply in group/community) and deletes it, preview with deleted message will be displayed in Activity center.

### ISSUE 4: Deleted message is visible in push notifications

If you think it's better to fix it in a separate PR, ping me pls and I'll log it.

### ISSUE 5: Redundant `!` button in the 'Message actions' context menu in Timeline

**_Reproduction:_**
1. Post smth in Timeline
2. Long tap on the message

Screenshot 2021-08-10 at 16 09 51

### ISSUE 6: Deleted messages in 1-1 and group chats are visible for sender in chat preview on Home screen

**_Reproduction:_**
1. Start 1-1 and create group chat with another user
2. Send messages to both chats
3. Delete messages
4. Check chat previews on Home screen

Screenshot 2021-08-10 at 16 41 26

**Comment by shivekkhurana** [comment]
**ISSUE 1** is a bug in Status-go
https://discord.com/channels/624205794384281629/862717376915308544/875284164931362877

**Comment by cammellos** [comment]
@qoqobolo issue 1 should be fixed in the next build, do you think is good enough to merge or there's some other issue that's blocking you think?

Thanks!

**Comment by qoqobolo** [comment]
@cammellos confirm that Issue 1 is fixed, that's great, thanks!

However, it would be nice to have fixed issues 3, 4, 5, and 6 in the PR if possible as well, since there are quite a few places left where the deleted messages remain visible for users.

But if you think it would be more convenient to fix it later, let me know, I'll log them separately and include them to 1.16.

**Comment by cammellos** [comment]
@qoqobolo if you could log them separately, it would be better, if that's ok, if we could merge, so we don't have to rebase, and I can take a look at the remaining issues.
Thanks!

**Comment by qoqobolo** [comment]
@cammellos sure, np!

**Pull request closed without merging**